buildbox: Add new cross-compiling buildbox for Teleport

camscale opened the pull request:
Add Dockerfiles and make targets to build a new buildbox that has a set
of cross compilers for the four architectures we target for Linux -
amd64, arm64, 386, and arm (32-bit). The buildbox also contains the
third-party C libraries that Teleport needs to statically link against
for a full set of features, again compiled for each target architecture.

The cross compilers are built using crosstool-NG. The compilers must
have a "vendor" field in the "target triple" of "unknown" (default).
Setting it to anything else means rust does not find the cross compilers
properly. Using "unknown" causes the triple to match the rust target.
(This may ultimately be unnecessary to make these match, but is needed
if we have the boring crate build boringssl itself - see last
paragraph).

Two container images are produced:
* `buildbox-thirdparty` - a base image that contains the third-party
  tools (compilers, etc) and libraries that we build from source. Once
  build, this should never need to be rebuilt unless we change the
  version of one of the components it builds.
* `buildbox-ng` - the buildbox used to build Teleport. It copies out the
  third-party components from the previous image and installs whatever
  other tools are needed by the build, whether from the distro archive
  or directly from upstream (such as Go and Rust compilers).

Additionally, the three intermediate build stages of
`buildbox-thirdparty` can be built for working on the buildbox; `ctng`,
`compilers` and `tplibs`.

`buildbox-ng` will become just `buildbox` once it can replace the others
entirely, at which point those others will be removed.

Currently the build tools to build a FIPS version of teleport is not in
the buildbox. That has been troublesome as the rust boring crate wants
to build boringssl itself and it is not being done correctly with the
cross compilers. Further work is needed here, likely building boringssl
ourselves and pointing the rust boring crate at the pre-built libraries.

---

This is still under development, but I want to get it out there for
visibility. Subsequent PRs will update the `build.assets/Makefile` to
build the teleport binaries, but there are some changes that need to be
made to the main Makefiles first so that both old and new buildboxes can
co-exist as well as being able to do native local builds from a teleport
checkout.

With this, you can run:

    make -C build.assets buildbox-thirdparty
    make -C build.assets buildbox-ng

and you'll get the final buildbox-ng image.

After doing that, you can also run

    make -C build.assets buildbox-thirdparty-ctng
    make -C build.assets buildbox-thirdparty-compilers
    make -C build.assets buildbox-thirdparty-tplibs

These will all come from the build cache and take almost no time at all.
These give you the intermediate stages of the `buildbox-thirdparty`
image so that you can reconfigure crosstool-NG for the different
architectures (see the `config/buildbox-ng` target in `Makefile`), and
enter at the different stages to run the compilers to build libraries.

camscale committed:
buildbox: Add new cross-compiling buildbox for Teleport

Add Dockerfiles and make targets to build a new buildbox that has a set
of cross compilers for the four architectures we target for Linux -
amd64, arm64, 386, and arm (32-bit). The buildbox also contains the
third-party C libraries that Teleport needs to statically link against
for a full set of features, again compiled for each target architecture.

The cross compilers are built using crosstool-NG. The compilers must
have a "vendor" field in the "target triple" of "unknown" (default).
Setting it to anything else means rust does not find the cross compilers
properly. Using "unknown" causes the triple to match the rust target.
(This may ultimately be unnecessary to make these match, but is needed
if we have the boring crate build boringssl itself - see last
paragraph).

Two container images are produced:
* `buildbox-thirdparty` - a base image that contains the third-party
  tools (compilers, etc) and libraries that we build from source. Once
  build, this should never need to be rebuilt unless we change the
  version of one of the components it builds.
* `buildbox-ng` - the buildbox used to build Teleport. It copies out the
  third-party components from the previous image and installs whatever
  other tools are needed by the build, whether from the distro archive
  or directly from upstream (such as Go and Rust compilers).

Additionally, the three intermediate build stages of
`buildbox-thirdparty` can be built for working on the buildbox; `ctng`,
`compilers` and `tplibs`.

`buildbox-ng` will become just `buildbox` once it can replace the others
entirely, at which point those others will be removed.

Currently the build tools to build a FIPS version of teleport is not in
the buildbox. That has been troublesome as the rust boring crate wants
to build boringssl itself and it is not being done correctly with the
cross compilers. Further work is needed here, likely building boringssl
ourselves and pointing the rust boring crate at the pre-built libraries.

jakule commented on build.assets/Makefile:
```diff
@@ -381,6 +439,58 @@ enter-root: buildbox
+	@: $(or $(wildcard buildbox/ct-ng-configs/$(CTNG_ARCH).defconfig),\
+		$(error No config for arch $(CTNG_ARCH)))
+	docker run -it --rm \
+		--hostname $(HOSTNAME) \
```
Why do we need hostname here?

camscale replied:
This is what was done with the other buildboxes: https://github.com/gravitational/teleport/blob/520539631dd973c2b3d4265250ad81803242bda4/build.assets/Makefile#L11

I can't use the `$(DOCKERFLAGS)` var as it has stuff I don't want so I just took the other bits. I'm ambivalent to setting hostname and am happy to remove this.

jakule replied:
If the docker need it I'd leave it and comment on why it's needed. If we can build without setting it I'd remove it.

camscale replied:
We don't need it. I think someone perhaps just liked it. I've removed it.

jakule commented on build.assets/Makefile:
```diff
@@ -381,6 +439,58 @@ enter-root: buildbox
+	docker run -it --rm \
+		--hostname $(HOSTNAME) \
+		--volume $(shell pwd)/..:/home/teleport \
+		--volume /tmp:/tmp \
```
Why do we need to mount /tmp? Can we mount subfolder instead of the whole /tmp?

camscale replied:
This is what is done for the other buildboxes. It doesn't matter on CI - it will always have a fresh /tmp for each job, but for local development, persisting the go and rust directories (GOPATH and CARGO_HOME which are set to /tmp directories for the buildbox) will make dev easier.

jakule replied:
In this case I'd use docker "internal caching" if you want to speed up the build. We're doing the same thing in TAG: https://github.com/gravitational/access-graph/blob/main/Dockerfile#L42
Otherwise you may hit weird filesystem related issues as "cannot remove root owned files" or similar.

camscale replied:
I'm using these caches for the download directory, but this cache is not for the `docker build` but for across `docker run` invocations, mostly to save some time during dev. It's not needed for functionality, so I've removed it for now and I'll look at what makes sense as an optimisation after.

jakule commented on build.assets/Makefile:
```diff
@@ -381,6 +439,58 @@ enter-root: buildbox
+	docker run -it --rm \
+		--hostname $(HOSTNAME) \
+		--volume $(shell pwd)/..:/home/teleport \
+		--volume /tmp:/tmp \
```
The same as above

jakule commented on build.assets/buildbox/bbcommon.mk:
nit: I'd rename the file bbc-ommon sounds a bit off IMO 😆 
What bb stands for?

camscale replied:
bb == buildbox. I didn't want to call it `common.mk` as there is already one of them at the top-level. I'll think about another name.

camscale replied:
I called it buildbox-common.mk

jakule reviewed (approved):
A few nits, but overall LGTM
@camscale Do you have the build time of Teleport with old/new toolchain? I'm curious how does it compare.

camscale committed:
buildbox: Add libelf.pc for pkg-config/pkconf

Add a libelf.pc file taken from elfutils and minorly adjusted for how we
install it separately. libelf was taken out of elfutils but did not take
the libelf.pc file, so we do that ourselves.

This will allow the build to use pkg-config/pkconf for selecting the
libraries to link teleport to as some later versions of libelf also need
`-lzstd`, such as this buildbox but also ubuntu-24.04.

camscale committed:
buildbox: Set prefix for zstd, add sh-cross-vars target

Set `PREFIX` when building `zstd` to ensure it's pkg-config file has the
correct prefix set in it, otherwise it has `-L/usr/local/lib` which we
do not want for cross-compiling.

Add a `sh-cross-vars` target to echo the cross-compiling vars for an
architecture, in a form that can be sourced by the shell:

    eval $(make -s -f cross-compile.mk ARCH=arm64 sh-cross-vars)

This will set up your shell for cross-compiling for the given
architecture, which is useful when working on the libraries in the
buildbox.

camscale commented on build.assets/buildbox/tplibs.mk:
I guess it is a bit confusing, since it stands for "third party" not "teleport". I'll call it `thirdparty-libs.mk`

camscale commented on build.assets/buildbox/tplibs.mk:
```diff
@@ -0,0 +1,235 @@
+include $(mk_dir)/bbcommon.mk
+include $(mk_dir)/cross-compile.mk
+
+# We build these libraries ourself and statically link them into the Teleport
+# binary as we need them build with PIE (Position Independent Executable) mode
+# so as to make use of ASLR (Address Space Layout Randomization). We cannot
+# rely on a host OS/packager to have built them this way.
```
To be honest, I don't really know. Since `-buildmode=pie` was added to the build in https://github.com/gravitational/teleport/pull/35543 I needed to modify the build Jakub originally had in his draft PR. Jakubs PR was prior to Reed's one just referenced so it all build fine. Once I rebased it to a later base, the libraries stopped building until I could figure out how to build each with PIE mode. That comment reflects that experience. Maybe the libraries in CentOS 7 are already build with PIE mode.

jakule replied:
> Why didn't we have to do this when building directly on CentOS 7?

This is not true. Reed rebuild all binaries on CentOS 7 to make them PIE compatible. If we want to build Teleport with PIE then all code must be build as PIC. As I mentioned, we can "reduce the complexity" by removing the libbpf from Teleport as OpenSSL by default builds as PIE and we don't really have any other C library that we need to build.

camscale replied:
There are a number of other C libraries we need to build: libtirpc, libpam, libudev_zero, libcbor, libfido2, libpcsclite.

But I noticed a difference between PIC and PIE. The libraries were already being build PIC, but that was not enough for PIE mode. When I started on this, I had to figure out how to build the libraries for `-buildmode=pie` as the existing PIC build was giving errors when linking.

camscale commented on build.assets/buildbox/tplibs.mk:
```diff
@@ -0,0 +1,235 @@
+openssl_TARGET_linux_amd64 = linux-x86_64
+openssl_TARGET_linux_arm64 = linux-aarch64
+openssl_TARGET_linux_386 = linux-x86
+#openssl_TARGET_linux_arm = linux-generic32
```
I'm not entirely sure which is more correct yet. I left it in as a reminder to me to figure it out.

camscale commented on build.assets/buildbox/ctng.mk:
Will do

camscale committed:
Address Roman's review comments

* Rename ctng to crosstoolng throughout
* Rename tplibs.mk to thirdparty-libs.mk
* Fix spelling mistakes.

camscale committed:
Address Jakub's review comments

* Remove `--hostname $(HOSTNAME)` when running containers.
* Remove `--volume /tmp:/tmp` when running containers.
* Rename bbcommon.mk to buildbox-common.mk

camscale commented:
@jakule > Do you have the build time of Teleport with old/new toolchain? 

I haven't done any timings yet. I am still having some trouble building enterprise teleport on arm64, which is the only platform I can reasonably compare at the moment with the rdp rust stuff (errors like `relocation truncated to fit: R_AARCH64_CALL26 against symbol `__aarch64_ldeor4_relax'`) when linking the final binary. I haven't noticed it being faster or slower in particular though.

@r0mant @jakule I've addressed all the comments. PTAL

jakule commented on build.assets/Makefile:
```diff
@@ -381,6 +439,52 @@ enter-root: buildbox
+.PHONY: config/buildbox-ng
+config/buildbox-ng:
+	@: $(or $(CROSSTOOLNG_ARCH),$(error CROSSTOOLNG_ARCH variable must be set))
+	@: $(or $(wildcard buildbox/ct-ng-configs/$(CROSSTOOLNG_ARCH).defconfig),\
```
```suggestion
	@: $(or $(wildcard buildbox/crosstoolng-config/$(CROSSTOOLNG_ARCH).defconfig),\
```

camscale replied:
Yeah, I'm sorry about that. I did the renaming suggested and stuffed up a few bits and wanted to get the PR updated before I went to sleep. I usually kick off a build before pushing, but these changes were going to take over an hour to build. I'll ensure I build before I push next time.

Anyway, fixed.

jakule commented on build.assets/buildbox/Dockerfile-thirdparty:
```diff
@@ -0,0 +1,131 @@
+USER buildbox
+WORKDIR /home/buildbox
+
+COPY cronsstoolng-configs cronsstoolng-configs
```
```suggestion
COPY crosstoolng-configs cronsstoolng-configs
```

jakule commented on build.assets/Makefile:
```diff
@@ -381,6 +439,52 @@ enter-root: buildbox
+	docker run -it --rm \
+		--volume $(shell pwd)/..:/home/teleport \
+		--workdir /home/teleport \
+		buildbox-thirdparty-crosstoolng \
```
This part breaks as the `buildbox-thirdparty-crosstoolng` container is tagged `teleportXX` where here we expect `latest` (default Docker behavior).

camscale replied:
Yeah, sorry, that was a side-effect of the renaming. I've fixed that now.

jakule commented on build.assets/Makefile:
```diff
@@ -107,6 +107,64 @@ build-binaries-fips: buildbox-centos7-fips webassets
+.PHONY: buildbox-ng
+buildbox-ng:
+	docker buildx build \
+		--build-arg THIRDPARTY_IMAGE=$(BUILDBOX_THIRDPARTY) \
```
This is tricky as it will always pick-up remote images instead of the local one. I had to modify this line to run it as the remove image doesn't exist.

camscale replied:
I have not had that problem, but I guess it has something to do with the builder? I think I might be using the original docker driver not the containerd one, but I'm not really sure what I'm seeing with docker once I had to switch to the Mac. I'm using colima, and spent a bit of time stuffing around with the runtime, snapshotter, etc but nothing made sense to me in the end.

I'm really not sure how I'm meant to set this stuff up if it will not look in the local image cache as it does for me but not you. Also in a way that works for devs on Macs and on Linux on CI.

jakule replied:
> I'm really not sure how I'm meant to set this stuff up

I'd just rename to local build to match the expected name here. In this case when you build locally the local container will be picked up instead of the remote one.

camscale replied:
I can't make that work. This is a "common" issue with the `docker-container` buildkit driver - it does not have access to the docker image cache except to store built images there when run with `--load`. I have been using the `docker` driver so this has not been an issue, but I ran the following and at the moment I cannot make this work with the `docker-container` driver:

```
$ docker buildx create foo --driver docker-container --use
$ docker images
ghcr.io/gravitational/teleport-buildbox-thirdparty   teleport17         bd51cf80045f   21 hours ago   2.08GB

# ... at this point I have thirdparty image. I give it a "local" name as you suggest:
$ docker tag ghcr.io/gravitational/teleport-buildbox-thirdparty:teleport17 teleport-buildbox-thirdparty:teleport17

# Now I try to build buildbox-ng with this image:
$ make -C build.assets buildbox-ng BUILDBOX_THIRDPARTY=teleport-buildbox-thirdparty:teleport17
...
Dockerfile:26
--------------------
  24 |     # Reference the thirdparty image for copying from later.
  25 |
  26 | >>> FROM ${THIRDPARTY_IMAGE} AS thirdparty
  27 |
  28 |     # ----------------------------------------------------------------------------
--------------------
ERROR: failed to solve: teleport-buildbox-thirdparty:teleport17: failed to resolve source metadata for docker.io/library/teleport-buildbox-thirdparty:teleport17: pull access denied, repository does not exist or may require authorization: server message: insufficient_scope: authorization failed
```

As you can see, it tries to resolve the "local" image name relative to `docker.io/library`.

I can't see any way to make this work using the `docker-container` driver. This is a long-standing issue (I dont want to link it here because I don't want that issue back-linking to this PR: moby buildkit issues/2343#issuecomment-1867944612) with no real resolution right now except to use the `docker` driver instead. But there is no standard name for the builder/context that uses that driver so it is not as simple as just telling docker buildx to use the `docker` driver.

Can you try building this using a build driver that uses the `docker` driver?

camscale replied:
I experimented with the containerd-snapshotter image cache, which can store multi-arch images, but it still doesn't work - it does not look in the local image cache at all when you are using the `docker-containerd` driver. @jakule If you did manage to get this working somehow, I'd really like to know what you did.

jakule commented on build.assets/buildbox/Dockerfile-thirdparty:
```diff
@@ -0,0 +1,131 @@
+
+# Build cross-compiling toolchains with ct-ng
+RUN --mount=type=cache,id=download,uid=${BUILDBOX_UID},target=${THIRDPARTY_DIR}/download \
+	make -f crosstoolng.mk crosstoolng ARCH=amd64 && \
```
This fails for me locally as:
* `crosstoolng` target doesn't exist
*  when I run `crosstoolng-menuconfig` (This is the correct one, kind of? technically we should run `make config` to create the config without any GUI) I'm getting permission denied.

jakule reviewed (commented):
I tried to run the build locally but it failed in a few places.

camscale committed:
buildbox: Fix up broken renaming

Renaming things was done poorly in the last few commits. Fix all that
up.

camscale committed:
buildbox: Use gold linker for arm64 builds

Use the `gold` linker when building arm64 binaries, as Enterprise
Teleport will not build with the binutils (bfd) linker; it gives
numerous errors when linking of the form:

    something.rs: (.text.unlikely._XXX): relocation truncated to fit: R_AARCH64_CALL26 against symbol ...

These errors do not occur when using the `gold` linker, which is already
included and built by crosstool-NG.

camscale committed:
build: Add release targets using buildbox-ng

Add a set of targets to build releases using buildbox-ng instead of the
standard buildboxes. Ultimately these targets will be removed when the
old buildboxes are removed. These new targets are temporary until that
happens supporting testing of the new buildbox.

camscale committed:
buildbox: Change crosstool arm tuple to match rust

Tweak the crosstool-NG configuration for the arm cross-compiler so that
the tuple for it matches the tuple currently used in the build by rust
for the same target. This is mostly to cause less confusion as there's
no real need for them to be different.

camscale committed:
buildbox: Add support for rust cross-compiling

Add environment variables so that cargo can find the appropriate
architecture-specific linker when cross-compiling. This is needed now as
we have a rust binary (fdpass-teleport) in the build, as opposed to just
rust compiled to a library linked to Go code. Rust does not have a cross
platform linker and relies an a linker in the toolchain for the target
architecture.

To make this work without needing per-architecture setup when building,
all of the toolchain binaries are symlinked into
/opt/thirdparty/host/bin so they are all in the path. Because each of
the binaries is prefixed with the target tuple, there are no name
clashes.

camscale committed:
buildbox: Add stages to buildbox Dockerfile for better caching

Download Go and Rust in separate stages and copy their installation into
the final container image. This helps as these stages no longer have
unrelated layers before them so they cache better. The Go and Rust
stages should only need to be rebuilt if the versions of the compilers
change, or some other aspect of the installation of the compilers
change. Previously adding a new package to be installed would cause Go
and Rust to be re-installed. This no longer happens.

camscale committed:
buildbox: Remove include of buildbox-common.mk in thirdparty-libs.mk

Remove the include of `buildbox-common.mk` from `thirdparty-libs.mk` as
it is included by `cross-compile.mk`. This caused some duplicate target
warnings when make ran.

camscale committed:
build: Add support for buildbox-ng to Makefile

Add supoprt for building in buildbox-ng to the build by including the
cross-compiling definitions if we are running in that buildbox and
gating some older cross-compiling definitions from being used.

camscale commented:
This is now working end-to-end for the releases of amd64, arm64, 386 and arm for OSS and enterprise. FIPS is still not supported though. The last thing I had to figure out was rust cross-compiling binaries - the rust compiler needs support of an external linker to cross-compile binaries and this is needed for the new `fdpass-teleport` binary. See [ba9a3bc](https://github.com/gravitational/teleport/pull/44130/commits/ba9a3bc3c8a2bfd6b8f79a0431bb10e134834430) for details.

I've also added some make targets to build.assets/Makefile to do release builds.

With this PR, you can build the buildbox and build the releases:

    $ cd build.assets
    $ make buildbox-thirdparty
    $ make buildbox-ng
    $ make release-ng-amd64
    $ ls -l ../build/artifacts  # <-- will show oss and enterprise tar balls
    $ make release-ng-arm64 # save the artifacts as each build does a clean
    $ make release-ng-386
    $ make release-ng-arm

It is possible that the second build (`make buildbox-ng`) will not work if you are using the containerd docker buildkit driver as it does not look in the local image cache for the image built with `make buildbox-thirdparty`. I am using the default docker driver and this works. I need to figure out what to do if you are using the containerd driver - perhaps it will all be fine once the images are pushed to ghcr.io, but it would be nice to not need that to help with local development.

camscale committed:
buildbox: Add teleport user, expand comments

Add the teleport user (1000:1000) to the buildbox for running in CI
where a repository is checked-out inside the container instead of
mounting a volume on to /home/teleport. Make that home directory
world-writable so that it can still be used if 1000:1000 cannot be used
for some reason.

Expand and clean up comments.

Re-order ENV vars a little for better grouping.

Put Rust and Go temp directories under /tmp/build so a single volume can
be mounted there for persistent caches across runs.

camscale committed:
build: Fix invocation of pkg-config in $(shell ...)

Explicitly set `PKG_CONFIG_PATH` when running `$(PKGCONF)` as when
running that in a `$(shell ...)` expression, variables exported in the
Makefile are not exported for that shell, so `$(PKGCONF)` does not see
`PKG_CONFIG_PATH`. GNU make 4.4 changes this so that exported variables
are exported to `$(shell ...)` expressions, but the buildbox has GNU
make 4.3.

This was causing libbpf to not be detected properly so was building
teleport without bpf.

camscale committed:
Merge branch 'master' into camh/buildbox-ng
